Fix omissions in GridGenerator::flatten_triangulation().

bangerth opened the pull request:
This is a final (?) follow-up to #13117, #13128, and related patches. It fixes the two bugs identified in #13126 and #13127, namely the omission of edges in 3d and the omission of copying manifold ids of internal faces and edges.

Fixes #13126 . Fixes #13127 .

/rebuild

bangerth commented:
This patch has #13142 as part of it. I will rebase once #13142 is merged, but the patch can be reviewed as is already.

masterleinad commented on source/grid/grid_generator.cc:
```diff
@@ -7538,41 +7540,187 @@ namespace GridGenerator
+            const_cast<Triangulation<dim, spacedim1> &>(in_tria)
+              .clear_user_flags_line();
+
+            // Loop over the boundary faces of the triangulation and create
```
```suggestion
            // Loop over all the faces of the triangulation and create
```

masterleinad commented on source/grid/grid_generator.cc:
```diff
@@ -7538,41 +7540,187 @@ namespace GridGenerator
+            const_cast<Triangulation<dim, spacedim1> &>(in_tria)
+              .clear_user_flags_quad();
+
+            // Loop over the boundary faces of the triangulation and create
```
```suggestion
            // Loop over all the faces of the triangulation and create
```

masterleinad commented on source/grid/grid_generator.cc:
```diff
@@ -7538,41 +7540,187 @@ namespace GridGenerator
+                      }
+
+                    // Then also loop over the edges of this face. Because every
+                    // edge must also be a part of a boundary face, we can
```
```suggestion
                    // boundary edge must also be a part of a boundary face, we can
```

masterleinad reviewed (approved):
Looks good to me.

bangerth commented:
#13142 is now merged, but the testers reveal that the MS VC compiler wants certain functions to be instantiated. That required me to work around an overload issue, so while I have dropped the commits that were part of #13142 and rebased, I needed to add two more commits to make things work. Let's see whether the MS VC compiler is happy.

bangerth commented:
It works now.

bangerth commented:
Ping?
